xpdan/callbacks_core.py

```diff
@@ -0,0 +1,206 @@
+
+        if not self.overwrite and os.path.isfile(filename):
+            raise OSError("There is already a file at {}. Delete "
+                          "it and try again.".format(filename))
+        if not self.dryrun:
+            tifffile.imsave(filename, np.asarray(image))
+            print("INFO: {} has been saved at {}"
+                  .format(fn, dir_path))
+
+        self.filenames.append(filename)
+
+    def _pull_dark_uid(self, doc, dark_field_key='sc_dk_field_uid'):
+        """method to relate dark to images
+
+        Simply replace this method if scheme is changed in the future
+        """
+        if 'dark_frame' in doc:
+            # found a dark header
+            dark_uid = None  # dark header, pass
+        else:
+            dark_uid = doc.get(dark_field_key, None)
+            if dark_uid is None:
+                print("INFO: no dark frame is associated in this header, "
+                      "subtraction will not be processed")
+                dark_uid = None  # can't find a dark
+        return dark_uid
+
+    def start(self, doc):
+        """method for start document"""
+        self.filenames = []
+        # Convert doc from dict into dottable dict, more convenient
+        # in Python format strings: doc.key == doc['key']
+        self._start = doct.Document('start', doc)
+
+        # find dark scan uid
+        dark_uid = self._pull_dark_uid(doc)
+        if dark_uid is None:
+            self.dark_img = None
+            self._find_dark = False
+        else:
+            dark_header = self.db[dark_uid]
+            self.dark_img = np.asarray(self.db.get_images(dark_header,
+                                                          self.image_field)
+                                      ).squeeze()
+            self._find_dark = True
+        super().start(doc)
+
+    def event(self, doc):
+        """tiff-saving operation applied at event level"""
+        if self.field not in doc['data']:
+            raise KeyError('required field = {} is not in header'
+                           .format(self.field))
+
+        self.db.fill_event(doc)  # modifies in place
+        image = np.asarray(doc['data'][self.field])
+
+        if self.dark_img is None:
+            # make a dummy one
+            self.dark_img = np.zeros_like(image)
+
+        if image.ndim == 2:
+            image = np.subtract(image, self.dark_img)
+            filename = self._generate_filename(doc)
+            path_dir, fn = os.path.split(filename)
+            if self._find_dark:
+                self._save_image(image, os.path.join(path_dir,
+                                                     'sub_'+fn))
+            else:
+                self._save_image(image, filename)
+            # if user wants wants raw dark
+            if self.save_dark:
+                self._save_image(self.dark_img, os.path.join(path_dir,
+                                                             'dark_'+fn))
+        if image.ndim == 3:
```

Comment:
Would it be possible to make an function which does this (if it doesn't exist already) and only operates on 2D arrays? That way if the array is 3D we can just call the 2D one a bunch of times? That might help us to avoid reusing code